File: 853-most-profit-assigning-work/most-profit-assigning-work.py
class Solution(object):
    def maxProfitAssignment(self, difficulty, profit, worker):
        jobs = sorted(zip(difficulty, profit))
        worker.sort()
        x = 0
        y = 0
        j = 0

        for i in worker:
            while j < len(jobs) and jobs[j][0] <= i:
                y = max(y, jobs[j][1])
                j += 1
            x += y
        
        return x

File: 853-most-profit-assigning-work/most-profit-assigning-work.java
import java.util.*;

class Solution {
    public int maxProfitAssignment(int[] difficulty, int[] profit, int[] worker) {
        List<Job> jobs = new ArrayList<>();
        for (int i = 0; i < difficulty.length; i++) {
            jobs.add(new Job(difficulty[i], profit[i]));
        }
        Collections.sort(jobs, (a, b) -> a.difficulty - b.difficulty);
        
        Arrays.sort(worker);
        
        int x = 0;
        int y = 0;
        int j = 0;
        
        for (int i = 0; i < worker.length; i++) {
            while (j < jobs.size() && jobs.get(j).difficulty <= worker[i]) {
                y = Math.max(y, jobs.get(j).profit);
                j++;
            }
            x += y;
        }
        
        return x;
    }
    
    private static class Job {
        int difficulty;
        int profit;
        
        Job(int difficulty, int profit) {
            this.difficulty = difficulty;
            this.profit = profit;
        }
    }
}

